--- Makefile.in	2003-06-12 16:31:31.000000000 +0200
+++ Makefile.in	2015-09-12 21:11:42.574797570 +0200
@@ -191,13 +191,13 @@
 	then echo $$r/m4/m4 ; \
 	else echo ${DEFAULT_M4} ; fi`
 
-# For an installed makeinfo, we require it to be from texinfo 4 or
-# higher, else we use the "missing" dummy.
+# For an installed makeinfo, we require it to be texinfo 4. texinfo >= 5 is more strict
+# and breaks the compilation. Disable makeinfo in that case by using the "missing" dummy.
 MAKEINFO=@MAKEINFO@
 USUAL_MAKEINFO = `if [ -f $$r/texinfo/makeinfo/makeinfo ] ; \
 	then echo $$r/texinfo/makeinfo/makeinfo ; \
 	else if (makeinfo --version \
-	  | egrep 'texinfo[^0-9]*([1-3][0-9]|[4-9])') >/dev/null 2>&1; \
+	  | egrep 'texinfo[^0-9]*([1-3][0-9]|[4])') >/dev/null 2>&1; \
         then echo makeinfo; else echo $$s/missing makeinfo; fi; fi`
 
 # This just becomes part of the MAKEINFO definition passed down to
