diff --git a/include/splay-tree.h b/include/splay-tree.h
index 86707fc..75a4d0b 100644
--- a/include/splay-tree.h
+++ b/include/splay-tree.h
@@ -44,8 +44,13 @@ extern "C" {
    these types, if necessary.  These types should be sufficiently wide
    that any pointer or scalar can be cast to these types, and then
    cast back, without loss of precision.  */
+#ifdef __MINGW64__
+typedef unsigned long long int splay_tree_key;
+typedef unsigned long long int splay_tree_value;
+#else
 typedef unsigned long int splay_tree_key;
 typedef unsigned long int splay_tree_value;
+#endif
 
 /* Forward declaration for a node in the tree.  */
 typedef struct splay_tree_node_s *splay_tree_node;
