--- gcc/Makefile.in	2018-12-22 20:33:32.805293600 +0200
+++ gcc/Makefile.in	2018-12-22 23:23:10.180206700 +0200
@@ -81,9 +81,9 @@
 # apply to the back end and the C front end, which may be compiled
 # with other compilers.  This is partially controlled by configure in
 # stage1, as not all versions of gcc understand -Wno-long-long.
-LOOSE_WARN = -W -Wall -Wwrite-strings -Wstrict-prototypes -Wmissing-prototypes
-STRICT_WARN = -Wtraditional @strict1_warn@
-STRICT2_WARN = -Wtraditional -pedantic -Wno-long-long
+LOOSE_WARN = -w
+STRICT_WARN = -w
+STRICT2_WARN = -w
 
 # This is how we control whether or not the additional warnings are applied.
 .-warn = $(STRICT_WARN)
